src/Servers/HttpSys/src/Helpers.cs

```diff
@@ -2,14 +2,35 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.Globalization;
+using System.Runtime.InteropServices;
 using System.Text;
 
 namespace Microsoft.AspNetCore.Server.HttpSys;
 
 internal static class Helpers
 {
-    internal static readonly byte[] ChunkTerminator = new byte[] { (byte)'0', (byte)'\r', (byte)'\n', (byte)'\r', (byte)'\n' };
-    internal static readonly byte[] CRLF = new byte[] { (byte)'\r', (byte)'\n' };
+    // Both of these arrays are pinned.
+    private static readonly byte[] ChunkTerminator = CreatePinnedArray("0\r\n\r\n"u8);
+    private static readonly byte[] CRLF = CreatePinnedArray("\r\n"u8);
+
+    private static byte[] CreatePinnedArray(ReadOnlySpan<byte> data)
+    {
+        var array = GC.AllocateArray<byte>(data.Length, pinned: true);
+        data.CopyTo(array);
+        return array;
+    }
+
+    internal static unsafe void GetCRLF(out byte* ptr, out int length)
+    {
+        ptr = (byte*)Marshal.UnsafeAddrOfPinnedArrayElement(CRLF, 0);
+        length = CRLF.Length;
+    }
+
+    internal static unsafe void GetChunkTerminator(out byte* ptr, out int length)
+    {
+        ptr = (byte*)Marshal.UnsafeAddrOfPinnedArrayElement(ChunkTerminator, 0);
+        length = ChunkTerminator.Length;
+    }
```

Comment:
> I would stay away from GC.AllocateArray(..., pinned: true). It is very hard to reason about its perf characteristics.

That makes me sad.

@davidfowl looks like we made the right choice ;)

Reply:
We are not going to change UTF8 string literal implementation to return moveable strings. It would be unnatural change and code out there depends on the UTF8 string literals to be pinned. I would be ok with simplifying it to 

```
internal static class Helpers
{
    // We are assuming that the UTF8 string literals are implicitly pinned
    // See https://github.com/dotnet/csharplang/issues/5295
    public static ReadOnlySpan<byte> ChunkTerminator => "0\r\n\r\n"u8;
    public static ReadOnlySpan<byte> CRLF => "\r\n"u8;
```

Reply:
The trouble with that is getting the pointer out of the underlying `ReadOnlySpan<T>`. I think for now I am going to leave this as is.

Reply:
`Unsafe.AsPointer(ref MemoryMarshal.GetReference(pinnedSpan))`. You need it in just one place here: https://github.com/dotnet/aspnetcore/pull/44561/files#diff-8279c4653ca902e3c8f3a3b68fef3fb12744e2b96cacc6f2215b86c712f750f7R298. (+change the method to take span and make the caller responsible for pinning the span).

Reply:
> Unsafe.AsPointer(ref MemoryMarshal.GetReference(pinnedSpan)). 

Yep, I was using that previously and have been trying to hide that implementation detail. Using the `fixed` keyword and then having it go out of scope is confusing in the best of times and in this case isn't clear this is permitted in this narrow instance. I also need it in https://github.com/dotnet/aspnetcore/pull/44561/files#diff-3e99422acb5996cdbb88be5ebbfa1d94e2c419f872da11f47f9fc5e90e835b43.